cpp/src/arrow/type.cc

```diff
@@ -3301,8 +3356,18 @@ std::shared_ptr<Field> field(std::string name, std::shared_ptr<DataType> type,
 }
 
 std::shared_ptr<DataType> decimal(int32_t precision, int32_t scale) {
-  return precision <= Decimal128Type::kMaxPrecision ? decimal128(precision, scale)
-                                                    : decimal256(precision, scale);
+  return precision <= Decimal32Type::kMaxPrecision    ? decimal32(precision, scale)
```

Comment:
Is it a breaking change? Users may get different data type and array type when precision is smaller than 18.

Reply:
This is a breaking change, and personally I'd prefer we deprecate this function.

Reply:
the updated version does exactly that, added a deprecation notice and created a new function